src/main/java/org/folio/circulation/services/LostItemFeeChargingService.java

```diff
@@ -0,0 +1,152 @@
+import static org.folio.circulation.support.ValidationErrorFailure.singleValidationError;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.folio.circulation.domain.AccountRepository;
+import org.folio.circulation.domain.FeeFine;
+import org.folio.circulation.domain.FeeFineOwner;
+import org.folio.circulation.domain.FeeFineOwnerRepository;
+import org.folio.circulation.domain.FeeFineRepository;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.policy.lostitem.LostItemPolicy;
+import org.folio.circulation.domain.policy.lostitem.ChargeAmountType;
+import org.folio.circulation.domain.policy.LostItemPolicyRepository;
+import org.folio.circulation.domain.representations.AccountStorageRepresentation;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.Result;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LostItemFeeChargingService {
+  private static final Logger log = LoggerFactory.getLogger(LostItemFeeChargingService.class);
+  private static final List<String> FEE_TYPES_TO_RETRIEVE = Arrays.asList(
+    LOST_ITEM_FEE_TYPE, LOST_ITEM_PROCESSING_FEE_TYPE);
+
+  private final LostItemPolicyRepository lostItemPolicyRepository;
+  private final FeeFineOwnerRepository feeFineOwnerRepository;
+  private final FeeFineRepository feeFineRepository;
+  private final AccountRepository accountRepository;
+
+  public LostItemFeeChargingService(Clients clients) {
+    this.lostItemPolicyRepository = new LostItemPolicyRepository(clients);
+    this.feeFineOwnerRepository = new FeeFineOwnerRepository(clients);
+    this.feeFineRepository = new FeeFineRepository(clients);
+    this.accountRepository = new AccountRepository(clients);
+  }
+
+  public CompletableFuture<Result<Loan>> chargeLostItemFees(Result<Loan> loanResult) {
+    return lostItemPolicyRepository.findLostItemPolicyForLoan(loanResult)
+      .thenCompose(loanWithPolicyResult -> loanWithPolicyResult.after(loan -> {
+        if (!shouldChargeAnyFee(loan.getLostItemPolicy())) {
+          log.debug("No fee is going to be charged, skipping logic, loan id {}", loan.getId());
+          return completedFuture(succeeded(loan));
+        }
+
+        return getFeeFineOwner(loan)
+          .thenCompose(r -> r.after(owner -> buildAccountsForPatron(owner, loan)))
+          .thenCompose(r -> r.after(accountRepository::createAll))
+          .thenApply(r -> r.map(notUsed -> loan));
+      }));
+  }
+
+  private CompletableFuture<Result<List<AccountStorageRepresentation>>> buildAccountsForPatron(
+    final FeeFineOwner owner, Loan loan) {
+
+    if (owner == null) {
+      log.warn("Can not find owner for service point {}, no fee will be charged",
```

Comment:
Could this log message include the service point and loan, to make it easier to act upon the feedback